src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3RequestStream.cs

```diff
@@ -0,0 +1,1277 @@
+                    case Http3FrameType.DuplicatePush:
+                    case Http3FrameType.PushPromise:
+                    case Http3FrameType.CancelPush:
+                        // Because we haven't sent any MAX_PUSH_ID frames, any of these push-related
+                        // frames that the server sends will have an out-of-range push ID.
+                        throw new Http3ConnectionException(Http3ErrorCode.IdError);
+                    default:
+                        // Unknown frame types should be skipped.
+                        await SkipUnknownPayloadAsync((Http3FrameType)frameType, payloadLength, cancellationToken).ConfigureAwait(false);
+                        break;
+                }
+            }
+        }
+
+        private async ValueTask ReadHeadersAsync(long headersLength, CancellationToken cancellationToken)
+        {
+            if (headersLength > _headerBudgetRemaining)
+            {
+                _stream.ResetSend((long)Http3ErrorCode.ExcessiveLoad);
+                throw new HttpRequestException(SR.Format(SR.net_http_response_headers_exceeded_length, _connection.Pool.Settings._maxResponseHeadersLength * 1024L));
+            }
+
+            _headerBudgetRemaining -= headersLength;
+
+            while (headersLength != 0)
+            {
+                if (_recvBuffer.ActiveLength == 0)
+                {
+                    _recvBuffer.EnsureAvailableSpace(1);
+
+                    int bytesRead = await _stream.ReadAsync(_recvBuffer.AvailableMemory, cancellationToken).ConfigureAwait(false);
+                    if (bytesRead != 0)
+                    {
+                        _recvBuffer.Commit(bytesRead);
+                    }
+                    else
+                    {
+                        throw new HttpRequestException(SR.net_http_invalid_response_premature_eof);
+                    }
+                }
+
+                int processLength = (int)Math.Min(headersLength, _recvBuffer.ActiveLength);
+
+                _headerDecoder.Decode(_recvBuffer.ActiveSpan.Slice(processLength), this);
+                _recvBuffer.Discard(processLength);
+                headersLength -= processLength;
+            }
+        }
+
+        private static ReadOnlySpan<byte> StatusHeaderNameBytes => new byte[] { (byte)'s', (byte)'t', (byte)'a', (byte)'t', (byte)'u', (byte)'s' };
+
+        void IHttpHeadersHandler.OnHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
+        {
+            Debug.Assert(name.Length > 0);
+            if (!HeaderDescriptor.TryGet(name, out HeaderDescriptor descriptor))
+            {
+                // Invalid header name
+                throw new HttpRequestException(SR.Format(SR.net_http_invalid_response_header_name, Encoding.ASCII.GetString(name)));
+            }
+            OnHeader(staticIndex: null, descriptor, staticValue: default, literalValue: value);
+        }
+
+        void IHttpHeadersHandler.OnStaticIndexedHeader(int index)
```

Comment:
Lets chat about this next week. I'm not as concerned about leaking some of QPack into H3 as they are tightly intertwined already, but I'd like to hear more about what other optimizations you feel we can get out of this abstraction.

Reply:
Initially I flagged this as a testing/maintenance issue. QPack can't be reviewed or tested in isolation because this chunk of it has been implemented directly inside the H3RequestStream.

Looking over this again I see a bigger problem. IHttpHeadersHandler is the boundary between the shared code (QPack) and the non-shared code (Http3RequestStream). All of the QPack logic you implement inside Http3RequestStream is code we have to duplicate in Kestrel. We should be able to share all of the QPack logic but to do so it needs to be factored appropriately.

Reply:
Follow-up on offline convo:

There's still concern that the abstraction as proposed will not provide real maintainability benefit and will be inefficient due to extra LUTs. We're going to give it a look post-merge to see if maintainability is improved.

There's a 2nd option we can keep in our back pocket of sharing type names between the two repos that would give the same level of abstraction without the added inefficiency.